src/main/java/com/vmware/vipclient/i18n/messages/api/opt/local/LocalLocaleOpt.java

```diff
@@ -37,72 +38,115 @@ public LocalLocaleOpt(LocaleDTO dto) {
+			cacheItem.set(System.currentTimeMillis());
+		}
     }
 
     private Map<String, String> getLanguagesNamesFromCLDR(String locale){
-		Map<String, String> languagesNames = null;
+		LocaleCacheItem languagesNames = null;
 		logger.debug("Look for languages' names from cache for locale [{}]!", locale);
 		FormattingCacheService formattingCacheService = new FormattingCacheService();
 		languagesNames = formattingCacheService.getLanguagesNames(locale);// key
 		if (languagesNames != null) {
+			if (languagesNames.isExpired()) { // cacheItem has expired
+				// Update the cache in a separate thread
+				populateLanguagesCache(locale, languagesNames);
+			}
 			logger.debug("Find languages' names from cache for locale [{}]!", locale);
-			return languagesNames;
+			return languagesNames.getCachedData();
 		}
-		languagesNames = getLanguagesNamesFromBundle(locale);
-		if (languagesNames != null) {
-			logger.debug("Find the languages' names from local bundle for locale [{}].\n", locale);// [datetime] and
-			formattingCacheService.addLanguagesNames(locale, languagesNames);
-			logger.debug("Languages' names are cached for locale [{}]!\n\n", locale);
-			return languagesNames;
-		}
-		return null;
+		languagesNames = new LocaleCacheItem();
+		getLanguagesNamesFromBundle(locale, languagesNames);
+		formattingCacheService.addLanguagesNames(locale, languagesNames);
+		logger.debug("Languages' names are cached for locale [{}]!\n\n", locale);
+		return languagesNames.getCachedData();
 	}
 
-	private JSONObject getLanguagesNamesFromBundle(String locale) {
+	private void getLanguagesNamesFromBundle(String locale, LocaleCacheItem cacheItem) {
 		logger.debug("Look for languages' names from local package bundle for locale [{}]!", locale);
 		String normalizedLocale = CommonUtil.getCLDRLocale(locale, localePathMap, localeAliasesMap);
 		logger.info("Normalized locale for locale [{}] is [{}]", locale, normalizedLocale);
 		if(normalizedLocale == null || normalizedLocale.isEmpty())
-			return null;
-        try {
-            String languagesJsonStr = PatternUtil.getLanguageFromLib(normalizedLocale);
-			JSONObject languagesData = (JSONObject) new JSONParser().parse(languagesJsonStr);
-            return (JSONObject) languagesData.get(PatternKeys.LANGUAGES);
+			return;
+		try {
+			String languagesJsonStr = PatternUtil.getLanguageFromLib(normalizedLocale);
+			Map<String, Object> languagesData = (Map<String, Object>) new JSONParser().parse(languagesJsonStr);
+			if (languagesData != null) {
+				logger.debug("Find the languages' names from local bundle for locale [{}].\n", locale);
+				cacheItem.set((Map<String, String>) languagesData.get(PatternKeys.LANGUAGES), System.currentTimeMillis());
+			}else{
+				logger.debug("Doesn't find the languages' names from local bundle for locale [{}].\n", locale);
+				cacheItem.set(System.currentTimeMillis());
```

Comment:
This is for handling non-supported locales, just want to add an empty fetch result to cache, it's the same as you new an empty MessageCacheItem and add it to cache.

Reply:
Same reply as in RemoteLocaleOpt and RemotePatternOpt:

_"The cacheItem object here was passed a parameter. It could be an object that was retrieved from the cache (not a new cacheItem), updating its timestamp here incorrectly "un-expires" the cacheItem. We want to keep the old timestamp if the fetch failed.

If you noticed in my most recent PR, we have changed the design of caching empty cacheItem objects for non-supported locales. This is because this design won't help in the case of the very first request (it will still be slow). So for L3, we just use the list of supported locales to figure out the next steps.. and in the case of non-supported locale, that would be to return fallback locale messages.

For L2, can we do the same in next PR? For now, remove this line please."_